doc/L3_performance_and_scaling_enchancements_HLD.md

```diff
@@ -0,0 +1,228 @@
+net.ipv4.neigh.default.gc_thresh2=32000
+net.ipv4.neigh.default.gc_thresh3=48000
+
+net.ipv6.neigh.default.gc_thresh1=8000
+net.ipv6.neigh.default.gc_thresh2=16000
+net.ipv6.neigh.default.gc_thresh3=32000
+```
+To increase rate of the ARP/ND packets coming to the CPU. Currently the max rate for ARP/ND is 600 packets, we will be increasing it to higher number(8000) in CoPP file  to improve the learning time.
+
+## 2.2 Performance Improvements
+
+This section elaborates the changes done to improve L3 performance in SONiC
+
+### 2.2.1 Route installation time
+
+The tables below captures the baseline route programming time for  IPv4 and IPv6 prefixes in SONiC.
+To measure route programming time, BGP routes were advertised to a SONiC router and timed how long it took for these routes to be installed in the ASIC.
+
+### Table 2: IPv4 prefix route programming time
+
+| Routes                  | time taken on AS7712(Tomahawk) |
+| ----------------------- | ------------------------------ |
+| 10k IPv4 prefix routes  | 11 seconds                     |
+| 30k IPv4 prefix routes  | 30 seconds                     |
+| 60k IPv4 prefix routes | 48 seconds                     |
+| 90k IPv4 prefix routes  | 68 seconds                     |
+
+
+
+### Table 3: IPv6 prefix route programming time
+
+| Routes                             | time taken on AS7712(Tomahawk) |
+| ---------------------------------- | -----------------------------  |
+| 10k IPv6 route with prefixes > 64b | 11 seconds                     |
+| 30k IPv6 route with prefixes >64b  | 30 seconds                     |
+
+
+#### 2.2.1.1 Proposed optimizations for reducing the route programming time.
+
+- <u>Using sairedis bulk route APIs</u>
+
+  In SONiC architecture, routeorch in Orchagent processes route table updates from the APP_DB and calls the sairedis APIs to put these routes in ASIC_DB. 
+  Currently Orchagent processes each route and puts in the ASIC_DB one at a time. The Redis pipelining allows for some level of bulking when putting entries in ASIC_DB but still one DB message is generated for every route.
+
+  Further bulking can be done by using the sairedis bulk APIs for route creation and deletion.
+
+  By using Sairedis bulk APIs, orchagent will call these APIs with a list of routes and their attributes.
+  The meta_sai layer in sairedis iterates over this route list and creates the meta objects for every route but only one Redis DB message will be generated for the route list. 
+  Therefore using the sairedis bulk APIs reduces the number of Redis messages. 
+  
+  The ASIC doesn't support bulk route creation/deletion,so syncd still processes one route at time and updates the ASIC.
+
+  So, the saving achieved by using bulk APIs will be number of Redis message generated.
+
+  Bulking of Route updates will be enabled in Orchagent. Orchagent will bulk 64 updates and send to Sairedis. 
+  A new timer will be introduced in orchagent to flush the outstanding updates every second.
+
```

Comment:
Bulking idea is nice but It introduces the delay for non bulk cases too. Can this timer tuned(starts off with low value and grow exponenetially, simiar to tcp)?